File: problems/37-bitmask-dp/002-traveling-salesperson/python/solution.py
def tsp(cost: list[list[int]]) -> int:
    n = len(cost)
    if n <= 1:
        return 0

    total_masks = 1 << n
    inf = 10**30
    dp = [[inf] * n for _ in range(total_masks)]
    dp[1][0] = 0

    for mask in range(1, total_masks):
        for last in range(n):
            if not (mask & (1 << last)) or dp[mask][last] == inf:
                continue

            for nxt in range(n):
                if mask & (1 << nxt):
                    continue

                next_mask = mask | (1 << nxt)
                candidate = dp[mask][last] + cost[last][nxt]
                dp[next_mask][nxt] = min(dp[next_mask][nxt], candidate)

    full_mask = total_masks - 1
    return min(dp[full_mask][last] + cost[last][0] for last in range(1, n))

File: problems/37-bitmask-dp/002-traveling-salesperson/java/Solution.java
import java.util.Arrays;

public class Solution {
    public int tsp(int[][] cost) {
        int n = cost.length;
        if (n <= 1) {
            return 0;
        }

        int totalMasks = 1 << n;
        long[][] dp = new long[totalMasks][n];
        long inf = Long.MAX_VALUE / 4;

        for (long[] row : dp) {
            Arrays.fill(row, inf);
        }

        dp[1][0] = 0; // Start at city 0.

        for (int mask = 1; mask < totalMasks; mask++) {
            for (int last = 0; last < n; last++) {
                if ((mask & (1 << last)) == 0 || dp[mask][last] == inf) {
                    continue;
                }

                for (int next = 0; next < n; next++) {
                    if ((mask & (1 << next)) != 0) {
                        continue;
                    }

                    int nextMask = mask | (1 << next);
                    long newCost = dp[mask][last] + cost[last][next];
                    dp[nextMask][next] = Math.min(dp[nextMask][next], newCost);
                }
            }
        }

        int fullMask = totalMasks - 1;
        long answer = inf;

        for (int last = 1; last < n; last++) {
            answer = Math.min(answer, dp[fullMask][last] + cost[last][0]);
        }

        return (int) answer;
    }
}
